Polys: Fixed the issue of cancel not expanding the product of irrational conjugates

iamabhishek0 opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests . Please also
write a comment on that issue linking back to this pull request once it is
open. -->
Fixes #18205 

#### Brief description of what is fixed or changed
--> In the issue # 18205, cancel((2 + I)*(2 - I)) was raising PolificationError
because in _parallel_dict_from_expr_no_gens expanded expression was passed,
which resulted in an empty Generator.

--> To fix this added I in gens, if the expression has either complex numerator or Denominator

--> To check whether the expression has complex Numerator or Denominator added a function is_complex

--> Added test in test_polytools
#### Other comments
After the fix
```
>>> cancel((2+I)*(3-I))
7 + I
>>> cancel((2+I)*(2-I))
5
```

#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* polys
  * Fixed the issue of cancel not expanding the product of irrational conjugates and added Tests


<!-- END RELEASE NOTES -->

v-shenoy committed:
Add numpy printing support for some matrix expressions

oscarbenjamin committed:
Add type hints in ciruitplot.py

oscarbenjamin committed:
Rename __import__kwargs to import_kwargs

oscarbenjamin committed:
Fix plotting imports in XFAIL tests

oscarbenjamin committed:
Ignore rubi in mypy

oscarbenjamin committed:
Ignore typing errors for external modules

oscarbenjamin committed:
Add type-hint in test_printing.py

oscarbenjamin committed:
Add type-hint for _argtypes in rv

oscarbenjamin committed:
Ignore typechecking in more modules

oscarbenjamin committed:
Move ImportErrors in circuitplot.py

oscarbenjamin committed:
Add eps argument to is_scalar_nonsparse_matrix

oscarbenjamin committed:
Use explicit methods for assumptions handlers

oscarbenjamin committed:
Make mypy ignore decorated property methods

oscarbenjamin committed:
Tell mypy to ignore missing stub files

oscarbenjamin committed:
Fix broken imports

oscarbenjamin committed:
Use tuples for __slots__

oscarbenjamin committed:
Use metaclass= instead of with_metaclass

oscarbenjamin committed:
Add type hints in multipledispatch

oscarbenjamin committed:
Remove unneeded imports in compatibility

oscarbenjamin committed:
Add type hint in logic

oscarbenjamin committed:
Add some type-hints in utilities

oscarbenjamin committed:
Add type hints in core

oscarbenjamin committed:
Re-add args methods in sets

oscarbenjamin committed:
Fix import error in circuitplot

oscarbenjamin committed:
Make pydy-example-repo not a package

oscarbenjamin committed:
Rewrite injection tests using exec

oscarbenjamin committed:
Add back sympify_mpq

oscarbenjamin committed:
Add type hints in polys

oscarbenjamin committed:
Add type hints in sets

oscarbenjamin committed:
Add type hints in tensor

oscarbenjamin committed:
Add type hints in core

oscarbenjamin committed:
Add type hints in diffgeom

oscarbenjamin committed:
Add type hints in _compilation

oscarbenjamin committed:
Include autolev .py and .al test files in install

oscarbenjamin committed:
Use type-comments for 3.5 in quantum and stats

oscarbenjamin committed:
Don't use TEST_SETUP with TEST_PY2_IMPORT

oscarbenjamin committed:
Don't test compileall under Python 2.7

oscarbenjamin committed:
Siloence mypy warnings in runtests.py

oscarbenjamin committed:
Fix condition in .travis.yml

oscarbenjamin committed:
Append sympy root directory in py2 import test

oscarbenjamin committed:
Ignore assign to method in runtests.py

oscarbenjamin committed:
Add type hints in utilities

oscarbenjamin committed:
Fix mypy errors in matrices

oscarbenjamin committed:
Use n instead of prec for evalf arg

oscarbenjamin committed:
Don't allow matrix subclasses to override simplify

oscarbenjamin committed:
Add type hints and fix mypy errors in matrices

oscarbenjamin committed:
Add type hints and fix mypy errors in vector

oscarbenjamin committed:
Remove redundant imports

oscarbenjamin committed:
Add type hints in printing

oscarbenjamin committed:
Add type hints in abc, assumptions, ordinal

oscarbenjamin committed:
Ignore sympy.benchmarks in mypy

oscarbenjamin committed:
Add type hints in codegen

oscarbenjamin committed:
Readd parameters in codegen

oscarbenjamin committed:
Fix handlers in assumptions

oscarbenjamin committed:
Remove unneeded import

smichr commented on sympy/polys/polyutils.py:
```diff
@@ -229,7 +229,7 @@ def _parallel_dict_from_expr_if_gens(exprs, opt):
 
 
-def _parallel_dict_from_expr_no_gens(exprs, opt):
+def _parallel_dict_from_expr_no_gens(exprs, opt,expression = None):
```
```suggestion
def _parallel_dict_from_expr_no_gens(exprs, opt, expression = None):
```

smichr commented on sympy/polys/polyutils.py:
```diff
@@ -274,6 +274,23 @@ def _is_coeff(factor):
 
+    complex_in_denom, complex_in_num = False, False
+    if not expression is None:
+        # is_complex returns complex_in_num = True if numerator is complex
```
the name `is_complex` is self documenting...no need for inlines here

smichr commented on sympy/polys/polyutils.py:
```diff
@@ -274,6 +274,23 @@ def _is_coeff(factor):
+    complex_in_denom, complex_in_num = False, False
+    if not expression is None:
+        # is_complex returns complex_in_num = True if numerator is complex
+        # and complex_in_denom = True if Denominator is complex
```
```suggestion
        # and complex_in_denom = True if denominator is complex
```

smichr commented on sympy/polys/polyutils.py:
```diff
@@ -321,7 +338,7 @@ def parallel_dict_from_expr(exprs, **args):
 
 
-def _parallel_dict_from_expr(exprs, opt):
+def _parallel_dict_from_expr(exprs, opt, expression = None):
```
```suggestion
def _parallel_dict_from_expr(exprs, opt, expression=None):
```

smichr commented on sympy/polys/polyutils.py:
```diff
@@ -421,6 +438,17 @@ def _dict_reorder(rep, gens, new_gens):
+def is_complex(exprs):
+    """Checks whether the numerator,Denominator of an expression is Complex"""
+    complex_in_num_denom = [False, False]
+    if len(exprs)>0:
```
```suggestion
    if len(exprs):
```

smichr commented on sympy/polys/polyutils.py:
```diff
@@ -421,6 +438,17 @@ def _dict_reorder(rep, gens, new_gens):
+    """Checks whether the numerator,Denominator of an expression is Complex"""
+    complex_in_num_denom = [False, False]
+    if len(exprs)>0:
+        if 'I' in str(exprs[0]):
```
This looks fragile...what if a variable has the name "I2"? Are you looking for something in a given form or something that has a non-zero imaginary component?

smichr replied:
Maybe you are just trying to target expressions `e` for which `pure_complex(e)` is not None:
```python
>>> from sympy.core.evalf import pure_complex
>>> pure_complex(3+I)
(3, 1)
>>> pure_complex(3)
```

iamabhishek0 replied:
I am just trying to check if the expression has any ImaginaryUnit in it
```
>>> from sympy.core.evalf import pure_complex
>>> pure_complex((2 + I)*(3 - I))
```
I was not expecting None here.
Can you please suggest any alternate way

smichr replied:
All expressions that have `I` in them? Then `exprs[0].has(I)` will work. But not all expressions that have I in them actually have imaginary components, e.g. roots of some cubics are returned in terms of `I` but are real.

What is the problem with `pure_complex` -- if you get None it means the expression does not look like `a + I*b`.

iamabhishek0 replied:
As discussed in #18205, There was inconsistency in the results because in the case of `cancel((2+I)*(2-I))` because PolificationFailed was raised https://github.com/sympy/sympy/blob/f1334bb2094a16bc5465dccbfac0acfbc7fb383f/sympy/polys/polytools.py#L4287 The cause of this was opt.gens being empty because in `def _parallel_dict_from_expr_no_gens(exprs, opt):` exprs that is passed is multiplied i.e 5 in case of (2 + I)*(2 - I)).
 So it is checking the actual expression and if it has `I` in it, `I` in added in gens https://github.com/sympy/sympy/blob/f1334bb2094a16bc5465dccbfac0acfbc7fb383f/sympy/polys/polyutils.py#L286 As you can see the coefficient of added term is 0. So in case of cubics roots won't get affected.

There's no problem with `pure_complex`, its just that I don't understand how it can be implemented in this case.

iamabhishek0 replied:
ping @smichr

iamabhishek0 commented:
Will update the PR with suggested changes.
Btw @smichr Thanks a lot for the review

oscarbenjamin committed:
Add type hints in assumptions and conftest.py

oscarbenjamin committed:
Add type hints in functions

oscarbenjamin committed:
Remove unneeded import

oscarbenjamin committed:
Add type hints in compatibility and geometry

oscarbenjamin committed:
Add type hints in integrals

oscarbenjamin committed:
Add type hints in parsing and solvers

oscarbenjamin committed:
Add type hints for SYMPY_DEBUG and in hyperexpand

oscarbenjamin committed:
Remove py3 annotation from sympy/__init__.py

oscarbenjamin committed:
Add type: ignore for multipledispatch

sachin-4099 committed:
Added required testcases

abh2k committed:
made suggested changes

abh2k committed:
changed name from general to polynomial

abh2k committed:
changed name from general to polynomial

abh2k committed:
minor changes

namannimmo10 committed:
Add bisectors method for Polygon class

sylee957 committed:
Merge pull request #18333 from sachin-4099/9471_wrong

Series: Added required testcases

Smit-create committed:
fixes 10076

abh2k committed:
made polynomial_congruence to accpt list of integers

abh2k committed:
minor changes

abh2k committed:
fixed is_nthroot

namannimmo10 committed:
Return sympy integers instead of python integers

namannimmo10 committed:
Add tests

czgdp1807 committed:
Merge branch 'master' into upright_latex

Smit-create committed:
Merge branch 'master' into 10076

namannimmo10 committed:
Add more tests to improve coverage

Smit-create committed:
added and corrected test

Smit-create committed:
unused imports

sylee957 committed:
Merge pull request #18326 from sachin-4099/16140_subs

Core: Fixed subs method which was not working with a partial derivative

namannimmo10 committed:
Further changes

tom-pytel committed:
squashed commits from 18313

tom-pytel committed:
flake8 fixes

jksuom committed:
Merge pull request #18337 from abhinav28071999/nth

fix is_nthpow_residue(a, n, m)

tom-pytel committed:
minor fix to adjugate and charpoly and tests

tom-pytel committed:
...

tom-pytel committed:
...

tom-pytel committed:
undoing commit to wrong branch

vikrantmalik051 committed:
Test Case added for diff(diff()) for Matrix

iamrajiv committed:
removed incorrect spacing with punctuation

czgdp1807 committed:
Merge pull request #18314 from Smit-create/issue2929

Added test of limit

namannimmo10 committed:
Update doctests

Upabjojr committed:
various fixes

oscarbenjamin committed:
Remove except Exception from codebase

oscarbenjamin committed:
Add type:ignore for multipledispatch in sets

oscarbenjamin committed:
Ignore pygletplot in mypy

sylee957 committed:
Merge pull request #18343 from V1krant/8207_calculus

Test Case added for diff(diff()) for Matrix Symbols

oscarbenjamin committed:
Add type-hint in testing

oscarbenjamin committed:
Add typehints in multipledispatch

oscarbenjamin committed:
Add type-hints in plotting

oscarbenjamin committed:
Add type:ignore for multipledispatch in matchpy

oscarbenjamin committed:
Fix type-hints in tensor

oscarbenjamin committed:
Add type-hints in stats

oscarbenjamin committed:
Add type-hints in parsing

namannimmo10 committed:
Suggestion from @sylee957

iamrajiv committed:
removed incorrect spacing with punctuation

iamrajiv committed:
removed incorrect spacing with punctuation

iamrajiv committed:
removed incorrect spacing with punctuation

sylee957 committed:
Use conjugate, transpose instead of getH, and match array shape of numpy.diag with sympy.DiagonalOf

jksuom committed:
Merge pull request #18202 from oscarbenjamin/pr_tmp_exceptions

Remove except Exception from codebase

oscarbenjamin committed:
Merge pull request #18351 from sylee957/numpy_matrix

Numpy printing support for some matrix expressions - revived

faizan2700 committed:
blank lines added

faizan2700 committed:
changes made

namannimmo10 committed:
Add test for polygon bisectors

namannimmo10 committed:
Another test for polygon with 5 vertices

asmeurer committed:
Merge pull request #18276 from Smit-create/17042

Using gmpy in integer_nthroot and igcd

gschintgen committed:
Fix FiniteSet union handler for symbolic elements

Fixes #18241.

Before:
In [1]: Interval(0, 1).union(FiniteSet(1, x))
Out[1]: [0, 1]

After:
In [1]: Interval(0, 1).union(FiniteSet(1, x))
Out[1]: [0, 1] ∪ {x}

Reason:
Incorrect handling of `.contains()` ternary logic
(True or False or Contains/Relational)

Tests are added.

abh2k committed:
improvd      description

oscarbenjamin committed:
Merge pull request #18356 from gschintgen/fix-18241-finiteset-union

Fix FiniteSet union handler for symbolic elements

Upabjojr committed:
fixing tests by sorting args in TensAdd

Upabjojr committed:
fixing test_pretty

oscarbenjamin committed:
Add type:ignore in sympy/physics/vector

oscarbenjamin committed:
Add type hints in units

oscarbenjamin committed:
Add type hints in units

oscarbenjamin committed:
Remove unneeded imports

Smit-create committed:
Suggested changes

czgdp1807 committed:
Merge pull request #14203 from TitanSnow/upright_latex

Use upright ``\LaTeX``

namannimmo10 committed:
Normalize so that distance between p1 and p2 is 1

sylee957 committed:
Merge pull request #18339 from namannimmo10/18338

Return sympy integers instead of python integers

abh2k committed:
improved      description

sylee957 committed:
Merge pull request #18335 from namannimmo10/geometry-new-feature

[feature] Add bisectors method for Polygon class

vikrantmalik051 committed:
Fixed Pretty Printing for Condition Set

iamrajiv committed:
updated URL

smichr committed:
Merge pull request #18345 from iamrajiv/pr_template

removed incorrect spacing with punctuation

vikrantmalik051 committed:
Test Case Fixed

vikrantmalik051 committed:
Hotfix

Commit:
Tensor: Added _eval_simplify method for class ImmutableDenseNDimArray.

--> _evalf_simplify simplifies each element of the array.
--> Added Tests

vikrantmalik051 committed:
Use Conditional Set in Tests

oscarbenjamin committed:
Add type-hints in physics

oscarbenjamin committed:
Merge pull request #18364 from V1krant/18272_printing

Fixed Pretty Printing for Condition Set

Commit:
returning self.applyfunc(simplify) in _eval_simplify

Smit-create committed:
Suggested change

Smit-create committed:
not rds

namannimmo10 committed:
Remove unreachable part of code

joha2 committed:
tensor: PartialDerivative: fixed inconsistent index generation in tensor derivative

The multiplication `*` seems to substitute all summation indices in an expression to dummies, which
is of no use if an expression is formed by generating each factor individually. Therefore, TensMul
was invoked directly, such that the expression does not get its dummy indices renamed before finalization.

Smit-create committed:
typo

Mohitbalwani26 committed:
Fix the trialset for undetermined coefficients in dsolve

oscarbenjamin committed:
Merge pull request #18309 from Mohitbalwani26/5096_bug

Fixed the way of creating trialset

oscarbenjamin committed:
Merge pull request #18244 from oscarbenjamin/pr_mypy_errors

[WIP] Fix some mypy errors to get mypy working

joha2 committed:
tensor: PartialDerivative: fixed another index error in tensor derivative calculation, added doit()

doit() is necessary to rename double indices in proper numbered dummys. Added some tests for
divergence type derivatives including consistency checks.

Mohitbalwani26 committed:
bug fix in checkodesol

Mohitbalwani26 committed:
tests added

abh2k committed:
added a function to check if expr is valid

abh2k committed:
added a check for coefficients to be integers

abh2k committed:
minor changes

abh2k committed:
minor changes

Smit-create committed:
fixing gmpy

Smit-create committed:
adding core tests in travis

Mohitbalwani26 committed:
removed commented code

Smit-create committed:
Merge branch 'master' into 10076

Commit:
Dropping Python 2: Removed Python 2 from translate()

oscarbenjamin committed:
Merge pull request #18376 from Mohitbalwani26/15996_bug

bug fix in checkodesol

Commit:
removed if py3

Commit:
removed import PY3

Smit-create committed:
removed deprecated

Smit-create committed:
unused import

Smit-create committed:
compile error

Smit-create committed:
blank line

oscarbenjamin committed:
Merge pull request #18385 from iamabhishek0/removed_python2_in_translate

Dropping Python 2: Removed Python 2 from translate()

Smit-create committed:
complie

Commit:
Removed Tests using Python 2

Commit:
removed unused imports

jksuom committed:
Merge pull request #18294 from abhinav28071999/general

Added general polynomial congruence

smichr committed:
Merge pull request #18373 from namannimmo10/remove-py2-

Remove unreachable part of code

Commit:
removed python2 compatible code

abh2k committed:
fixed bug

Commit:
removed _rint with int

czgdp1807 committed:
Merge pull request #18336 from Smit-create/10076

Fixed Wrong Integral from SingleContinuousPSpace

Smit-create committed:
changing 2**63

Smit-create committed:
trailing space

sylee957 committed:
Merge pull request #18369 from iamabhishek0/_evalf_simplify_#18361

Tensor: Added _eval_simplify method for class ImmutableDenseNDimArray.

Smit-create committed:
fixed limit

jksuom committed:
Merge pull request #18406 from Smit-create/18306

Fixed NotImplemeted Error in limit

namannimmo10 committed:
Fix typo and add tests related to gcdex_diophantine

tom-pytel committed:
removed everything except function relocation

tom-pytel committed:
flake8 fixes

tom-pytel committed:
fixed docstring reference

Mohitbalwani26 committed:
added hyperbolic trignometric in ode

Mohitbalwani26 committed:
synching

oscarbenjamin committed:
Convert ode module into a package

oscarbenjamin committed:
Fix imports after moving the ode module

oscarbenjamin committed:
Fix imports after moving the ode module

oscarbenjamin committed:
Merge pull request #18393 from iamabhishek0/removing_python_2

Dropping Python 2: Removed Python 2 compatibility

Upabjojr committed:
Merge pull request #18093 from joha2/18048_tensor_derivatives

tensor: Implementation of tensorial and symbolic partial derivatives

mijo2 committed:
core: Added new test case for verifying if replace method works with Equality expression

Added a new test case to check if the replace method exhibits correct behaviour with
Equality and in turn Relational. There was a bug in sympy 1.5.1 where the test case
mentioned below didn't work. Even though it was resolved due to a fix of some other
issue, this test case will still prove necessary for future versions to get the desired
results when replace and Equality/Relational are used in conjunction.

Example:

>> import sympy as sp
>> x, y, z = sp.symbols("x, y, z")
>> Xeq = sp.Eq(x, sp.pi * y**3 * (sp.Rational(1, 6) + z / 4 / y))
>> Xeq.replace(y**3, z)x
Eq(x, pi*z*(1/6 + z/(4*y)))

sylee957 committed:
Merge pull request #18342 from Pristine-Cat/18313_funcs

Pulled MatrixDeterminant function implementations out of matrix API

Mohitbalwani26 committed:
added hint in tests

sylee957 committed:
Fix LaTeX printing of Range

Smit-create committed:
Suggested changes

Mohitbalwani26 committed:
added hint in tests

namannimmo10 committed:
Suggestions from @V1krant

namannimmo10 committed:
Merge remote-tracking branch 'upstream/master' into new_patch

tom-pytel committed:
reductions and subspaces

tom-pytel committed:
flake8 fixes

Commit:
Removed python2 compatible code

oscarbenjamin committed:
Merge pull request #18416 from mijo2/18412_new_test_case

Addition of a test to verify if replace works with Equality/Relational or not

oscarbenjamin committed:
Merge pull request #18390 from Smit-create/ask_dep

Removed deprecated in assumptions

oscarbenjamin committed:
Merge pull request #18418 from sylee957/fix_latex_range

Fix LaTeX printing of Range containing symbols

tom-pytel committed:
fixed MatrixSubspaces.orthogonalize docstring

oscarbenjamin committed:
Merge pull request #18422 from iamabhishek0/removing_python_2

Dropping Python 2: Removed python2 compatible code

oscarbenjamin committed:
Merge pull request #18410 from Mohitbalwani26/18408_bug

Added hyperbolic function in undetermined coefficients

abh2k committed:
minor improvements

mohitacecode committed:
Change the logic of the eval method

sylee957 committed:
Merge pull request #18420 from Pristine-Cat/18313_funcs2

Pulled MatrixReductions and MatrixSubspaces function implementations out of Matrix API

mohitacecode committed:
Update integers.py

abh2k committed:
minor changes

tom-pytel committed:
eigen

oscarbenjamin committed:
Update doc references after moving ode module

namannimmo10 committed:
Suggestions from @smichr and @jksuom

gschintgen committed:
Reject more non-real intervals, e.g. Interval(x, x + I)

gschintgen committed:
Fix handling of (extended) reals in Interval._contains

Sympy `Interval`s are purely real and hence `._contains()`
rejects e.g. S.Infinity. This commit fixes the following two
results that were not quite consistent with this definition:

In [1]: c = Symbol('c', real=False)
In [2]: Interval(m, m + 1).contains(c)
Out[2]: c ≥ m ∧ c ≤ m + 1  # should be False

In [3]: e = Symbol('e', extended_real=True)
In [4]: Interval(-oo, oo).contains(e)
Out[4]: True  # is in fact unknown!

mohitacecode committed:
Added Tests for issue 18421

oscarbenjamin committed:
Merge pull request #18192 from namannimmo10/new_patch

Handle Mul.is_imaginary for infinite values

oscarbenjamin committed:
Fix import after moving ode module

oscarbenjamin committed:
Add __init__.py in sympy.solvers.ode.tests

smichr committed:
Merge pull request #18409 from namannimmo10/14704

Fix typo and add tests related to gcdex_diophantine

abh2k committed:
added test cases

abh2k committed:
minor changes

mohitacecode committed:
Update test_integers.py

mohitacecode committed:
Update test_integers.py

abh2k committed:
removed white trailing spaces

mohitacecode committed:
Update test_integers.py

smichr committed:
Merge pull request #18398 from abhinav28071999/doc

fixed is_nthpow_residue

mohitacecode committed:
Update test_integers.py

gschintgen committed:
Intervals: implement review suggestions

oscarbenjamin committed:
Merge pull request #18426 from gschintgen/sets-fixes-v2

Minor Interval fixes

oscarbenjamin committed:
Merge pull request #18425 from Pristine-Cat/18313_funcs3

Pulled MatrixEigen function implementations out of Matrix API

oscarbenjamin committed:
Merge pull request #18380 from Smit-create/gmpy

Fixing gmpy in integer_nthroot

oscarbenjamin committed:
Merge pull request #18317 from sylee957/improve_lu_docs

Improve docs for LUdecomposition_Simple

oscarbenjamin commented:
This needs a review

oscarbenjamin committed:
Merge pull request #18267 from faizan2700/master

test for Ellipse.evolute and Ellipse.second_moment_area added

smichr commented on sympy/polys/polyutils.py:
```diff
@@ -273,6 +273,23 @@ def _is_coeff(factor):
+    I = ImaginaryUnit()
+
+    # if expression is imaginary but gens is empty then add I in gens
+    if not I in gens and len(gens) == 0 and (imag_in_num or imag_in_denom):
```
If len(gens) is 0 then I can't be in gens so isn't that a redundant condition?

smichr commented on sympy/polys/polyutils.py:
```diff
@@ -420,6 +437,28 @@ def _dict_reorder(rep, gens, new_gens):
     return map(tuple, new_monoms), coeffs
 
+def has_imaginary(exprs):
+    """Checks whether the numerator, denominator of an expression has Imaginary Unit"""
```
This is not telling what form `exprs` is in. It says "expression" but then you check for `len(exprs)` below which suggests that it is an iterable.

smichr commented on sympy/polys/polyutils.py:
```diff
@@ -420,6 +437,28 @@ def _dict_reorder(rep, gens, new_gens):
 
+def has_imaginary(exprs):
+    """Checks whether the numerator, denominator of an expression has Imaginary Unit"""
+    has_imag_in_num_denom = [False, False]
```
I would suggest making this `in_num, in_den = [False]*2` and return `in_num, in_den` at the end rather than use an index which will force someone to look back to make sure that the order is num/den (though I see you use the name to give that hint). Still...I would prefer to see direct rather than indirect referencing.

smichr commented:
My hunch is this is overcomplicating the issue. Even the non-I case generates the error:
```python
>>> Poly(5,x)
Poly(5, x, domain='ZZ')
>>> cancel(_)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "sympy\polys\polytools.py", line 6614, in cancel
    f = factor_terms(f, radical=True)
  File "sympy\core\exprtools.py", line 1256, in factor_terms
    return do(expr)
  File "sympy\core\exprtools.py", line 1252, in do
    *[do(a) for a in p.args])
  File "sympy\polys\polytools.py", line 125, in __new__
    return cls._from_expr(rep, opt)
  File "sympy\polys\polytools.py", line 236, in _from_expr

    return cls._from_dict(rep, opt)
  File "sympy\polys\polytools.py", line 174, in _from_dict

    "can't initialize from 'dict' without generators")
sympy.polys.polyerrors.GeneratorsNeeded: can't initialize from 'dict' without generators
```

smichr commented:
I am not sure what the ramifications are. Perhaps @jksuom could advise: can we just give generator `x` to any Poly created from an integer?
```diff
diff --git a/sympy/polys/polyutils.py b/sympy/polys/polyutils.py
index 89d04c5..c8f838c 100644
--- a/sympy/polys/polyutils.py
+++ b/sympy/polys/polyutils.py
@@ -3,7 +3,7 @@
 from __future__ import print_function, division
 
 from sympy.core import (S, Add, Mul, Pow, Eq, Expr,
-    expand_mul, expand_multinomial)
+    expand_mul, expand_multinomial, symbols)
 from sympy.core.exprtools import decompose_power, decompose_power_rat
 from sympy.polys.polyerrors import PolynomialError, GeneratorsError
 from sympy.polys.polyoptions import build_options
@@ -298,7 +298,9 @@ def _is_coeff(factor):
                 poly[monom] = Mul(*coeff)
 
         polys.append(poly)
-
+    if not gens:
+        gens = symbols('x,')
+        polys = [{(0,): polys[i][()]} for i in range(len(polys))]
     return polys, tuple(gens)
 ```
 ```python
>>> Poly(3)
Poly(3, x, domain='ZZ')
>>> Poly(3+I)
Poly(I + 3, I, domain='ZZ')
>>> cancel((2+I)*(2-I))
5
>>> cancel((2+I)*(3-I))
7 + I
```

mohitacecode committed:
Fixes line spacing

smichr commented:
Or maybe an "invisible" symbol could be used:
```python
>>> Poly(3,Symbol(''))
Poly(3, , domain='ZZ')
```

jksuom commented:
I think that it should be possible to solve this without having to modify the low level methods `_parallel_...`. If there are no generators, then I would expand `f` before returning it.

oscarbenjamin committed:
Merge pull request #18424 from mohitacecode/patch-5

Change the logic of the eval method present in RoundFunction Class (integers.py)

oscarbenjamin committed:
Make constantsimp importable from ode

oscarbenjamin committed:
Merge pull request #18414 from oscarbenjamin/pr_move_ode

Convert the ODE module into a package

iamabhishek0 commented:
@jksuom @smichr I tried the above-suggested solution but it this throwing exception in most of the tests in `test_polytools.py`

jksuom commented:
I would first try something simple like this:
```
--- a/sympy/polys/polytools.py
+++ b/sympy/polys/polytools.py
@@ -6625,7 +6625,7 @@ def cancel(f, *gens, **args):
         (F, G), opt = parallel_poly_from_expr((p, q), *gens, **args)
     except PolificationFailed:
         if not isinstance(f, (tuple, Tuple)):
-            return f
+            return f.expand()
         else:
             return S.One, p, q
     except PolynomialError as msg:
```

Commit:
Merge remote-tracking branch 'github/18205_irrational_conjugates' into 18205_irrational_conjugates

Commit:
reverted changes

Commit:
return f.expand() if PolificationFailed

iamabhishek0 commented:
@smichr I am extremely sorry for the bad commit

iamabhishek0 commented:
I am opening a New PR, If thats not a problem
